open-apparel-registry/open-apparel-registry
# Store raw data

## Overview

When there is an error in list processing, it's sometimes unclear whether that error was caused by an improperly formatted list file or by a bug in the code. Debugging is much more efficient and effective when we have access to the originally uploaded file. We now store the raw data file in S3 when it's uploaded, and enable downloading via both the list details and the Django admin. 

Connects #2119 

## Demo

<img width="1428" alt="Screen Shot 2022-09-21 at 9 08 36 AM" src="https://user-images.githubusercontent.com/21046714/191512755-ca9a2384-817f-4c9a-ad92-6e3198143e27.png">
<img width="1083" alt="Screen Shot 2022-09-21 at 9 09 15 AM" src="https://user-images.githubusercontent.com/21046714/191512774-8583b8b6-6e26-4fed-a8db-54560fb9f400.png">

## Testing Instructions

* Run `./scripts/update` and `./scripts/server`
* Contribute a file. It should save the the local file system. (Note: The download buttons won't work.)
* In `docker-compose.yml`, set the `AWS_STORAGE_BUCKET_NAME` to `opensupplyhub-testing-files-eu-west-1`.
* In `settings.py`, change the conditional `if not DEBUG` on line 462 to be True (to cause Django to use S3 instead of the local filesystem)
* Contribute a file. Upload should succeed. 
* Visit the [opensupplyhub-testing-files-eu-west-1](https://s3.console.aws.amazon.com/s3/buckets/opensupplyhub-testing-files-eu-west-1?region=eu-west-1&tab=objects) bucket on S3. The file you uploaded should be there. 
* On the list page, use the 'download submitted csv' button to download the CSV.
* On the Django admin, use the download button in the FacilityList's details to download the CSV. 
* Contribute a file with the same name. View the bucket on S3 and confirm that the original file wasn't overwritten. 
* You should be able to download the second CSV via the list details and Django admin. 
* Visit a list details page that was uploaded on a different branch (ie where the list file hasn't been stored). The list should load without error, and the button for the original file should be hidden. 

## Checklist
